libs/extended/src/System/Logger/Extended.hs

```diff
@@ -1,32 +1,90 @@
+{-# LANGUAGE DeriveAnyClass #-}
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
-{-# LANGUAGE DerivingStrategies         #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
 -- | Tinylog convenience things.
 module System.Logger.Extended
     ( module Log
+    , LogFormat(..)
     , mkLogger
     , mkLogger'
     , LoggerT(..)
     , runWithLogger
+    , netStringsToLogFormat
     ) where
 
-import Imports
 import Control.Monad.Catch
+import Data.Aeson
+import Data.Aeson.Encoding (pair, list, text)
 import Database.CQL.IO
+import GHC.Generics
+import Imports
 import System.Logger as Log
-
-import qualified Data.ByteString.Lazy.Char8 as L
+import Data.String.Conversions (cs)
 import qualified Data.ByteString.Lazy.Builder as B
+import qualified Data.ByteString.Lazy.Char8 as L
 import qualified System.Logger.Class as LC
 
--- TODO(arianvp): Get rid of boolean blindness
--- TODO(arianvp): Add JSON log format. This will make our lives a lot easier
---    This will add a dependency on aeson for this package,
---    but it already transitively depended on it through imports.
---    Interestingly, the only place where imports uses Aeson
---    is in the Orphans module which defines Aeson Orphans instances for Log.Level.
---    So while we're at it, we should probably move those orphans here.
-mkLogger :: Log.Level -> Bool -> IO Log.Logger
-mkLogger lvl netstr = Log.new
+
+deriving instance Generic LC.Level
+instance FromJSON LC.Level
+instance ToJSON LC.Level
+
+-- | The log formats supported
+data LogFormat = JSON | Plain | Netstring 
+  deriving stock (Eq, Show, Generic)
+  deriving anyclass (ToJSON, FromJSON) -- TODO write this instance manually?
+
+-- | We use this as an intermediate structure to ease the implementation of the
+-- ToJSON instance but we could just inline everything. I think this has
+-- negligible impact and makes the code a bit more readable. Let me know
+data Element' = Element' [(Builder, Builder)] [Builder]
+
+instance ToJSON Element' where
+  toJSON = error "Use toEncoding"
```

Comment:
https://hoogle.zinfra.io/file/root/.stack/snapshots/x86_64-linux/custom-wire-1.1-G6NVtOVxqIB2/8.4.3/doc/aeson-1.3.1.1/Data-Aeson.html#t:ToJSON

> Instances in general must specify toJSON and should (but don't need to) specify toEncoding.

What is the reason for implementing `toEncoding`, and not `doJSON`?

Reply:
Yeh I could, but this `Element'` datatype is private to this module, and we only call `toEncoding` on it and not `toJSON`  . My original plan was to get rid of the intermediate `Element'` datatype completely so that no intermediate structure is allocated at all and it would be streaming bytestrings all the way down.

I'll just get rid of the `ToJSON` instance and rename `toEncoding` to `toElementEncoding`